deeppavlov/models/go_bot/trippy.py

```diff
@@ -0,0 +1,479 @@
+            db_slots = {
+                s: v for s, v in self.ds.items() if v != 'dontcare' and s in self.database.keys
+            }
+
+            db_results = self.database([db_slots])[0]
+
+            # filter api results if there are more than one
+            # TODO: add sufficient criteria for database results ranking
+            if len(db_results) > 1:
+                db_results = [r for r in db_results if r != self.db_result]
+            else:
+                print("Failed to get any results for: ", db_slots)
+        else:
+            logger.warning("No database specified.")
+
+        logger.info(f"Made api_call with {self.ds.keys()}, got {len(db_results)} results.")
+        self.current_db_result = {} if not db_results else db_results[0]
+        self._update_db_result()
+
+    def _update_db_result(self):
+        if self.current_db_result is not None:
+            self.db_result = self.current_db_result
+
+    def update_ground_truth_db_result_from_context(self, context: Dict[str, Any]) -> None:
+        self.current_db_result = context.get('db_result', None)
+        self._update_db_result()
+
+    def fill_current_state_with_db_results(self) -> None:
+        if self.db_result:
+            for k, v in self.db_result.items():
+                self.ds[k] = str(v)
+
+    def train_on_batch(self,
+                       batch_dialogues_utterances_features: List[List[dict]],
+                       batch_dialogues_utterances_targets: List[List[dict]]) -> dict:
+        """
+        Train model on given batch.
+
+        Args:
+            batch_dialogues_utterances_features:
+            batch_dialogues_utterances_targets: 
+
+        Returns:
+            dict with loss value
+        """
+        # Turns on dropout
+        self.model.train()
+        # Zeroes grads
+        self.model.zero_grad()
+        batch, features = prepare_trippy_data(batch_dialogues_utterances_features,
+                                              batch_dialogues_utterances_targets,
+                                              self.tokenizer,
+                                              self.slot_names,
+                                              self.class_types,
+                                              self.nlg_manager,
+                                              self.max_seq_length,
+                                              debug=self.debug)
+        # Move to correct device
+        batch = batch_to_device(batch, self.device)
+
+        if not(self.slot_names):
+            batch["start_pos"] = None
+            batch["end_pos"] = None
+            batch["inform_slot_id"] = None
+            batch["refer_id"] = None
+            batch["class_label_id"] = None
+            batch["diag_state"] = None
+
+        # Feed through model
+        outputs = self.model(**batch)
+
+        # Backpropagation
+        loss = outputs[0]
+        action_loss = outputs[7]
```

Comment:
Ah actually I just realized it's better to keep it separated as we're adding the huggingface outputs, i.e.
`outputs = (total_loss,) + (per_slot_per_example_loss, per_slot_class_logits, per_slot_start_logits, per_slot_end_logits, per_slot_refer_logits, action_logits, action_loss,) + outputs[2:]`

So depending on the huggingface version it may vary how long outputs[2:] is..